File: frontend/algorithmic-trading-frontend/src/services/predictions.ts
import { RealtimePrediction, HistoricalPrediction } from '@/components/predictions/usePredictions';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/predictions';

interface ApiResponse<T> {
  status: 'success' | 'error';
  data?: T;
  message?: string;
}

/**
 * Fetches the latest real-time prediction for a given symbol.
 * @param symbol The stock symbol to fetch.
 * @returns A promise that resolves to the real-time prediction data.
 */
export const getRealtimePrediction = async (symbol: string): Promise<RealtimePrediction> => {
  try {
    const response = await fetch(`${API_BASE_URL}/realtime/${symbol}`);
    if (!response.ok) {
      throw new Error(`API call failed with status: ${response.status}`);
    }
    const result: ApiResponse<RealtimePrediction> = await response.json();
    if (result.status === 'error' || !result.data) {
      throw new Error(result.message || 'Failed to fetch real-time prediction.');
    }
    return result.data;
  } catch (error) {
    console.error(`Error fetching real-time prediction for ${symbol}:`, error);
    // Return a mock error response or rethrow
    throw error;
  }
};

/**
 * Fetches historical prediction data for a given symbol.
 * @param symbol The stock symbol to fetch.
 * @param limit The number of data points to retrieve.
 * @returns A promise that resolves to an array of historical predictions.
 */
export const getHistoricalPredictions = async (symbol: string, limit: number = 100): Promise<HistoricalPrediction[]> => {
  try {
    const response = await fetch(`${API_BASE_URL}/history/${symbol}?limit=${limit}`);
    if (!response.ok) {
      throw new Error(`API call failed with status: ${response.status}`);
    }
    const result: ApiResponse<HistoricalPrediction[]> = await response.json();
    if (result.status === 'error' || !result.data) {
      throw new Error(result.message || 'Failed to fetch historical predictions.');
    }
    return result.data;
  } catch (error) {
    console.error(`Error fetching historical predictions for ${symbol}:`, error);
    throw error;
  }
};


/**
 * Fetches the model's accuracy metrics.
 * @returns A promise that resolves to the accuracy data.
 */
export const getModelAccuracy = async (): Promise<any> => {
    try {
      const response = await fetch(`${API_BASE_URL}/accuracy`);
      if (!response.ok) {
        throw new Error(`API call failed with status: ${response.status}`);
      }
      const result: ApiResponse<any> = await response.json();
      if (result.status === 'error' || !result.data) {
        throw new Error(result.message || 'Failed to fetch model accuracy.');
      }
      return result.data;
    } catch (error) {
      console.error('Error fetching model accuracy:', error);
      throw error;
    }
};

// WebSocket logic is handled by the usePredictions hook, 
// but this service could be expanded to include more complex API interactions.

File: nautilus_trader_engine/api/routers/predictions.py
"""
API Router for Real-Time Trading Predictions

This module provides API endpoints for accessing real-time and historical
trading predictions, including a WebSocket endpoint for live streaming.

Features:
- WebSocket endpoint for streaming real-time predictions.
- REST endpoints for historical prediction data and model accuracy.
- Prediction caching and performance optimization.
- Prediction confidence scoring and validation.

Author: Vincent Pereira
Version: 2.0.0
"""

import asyncio
import json
from datetime import datetime, timezone
from typing import List, Dict, Any

import websockets
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, Path, Query, Depends, HTTPException, status
from caching.cache_client import CacheClient, get_cache_client

from nautilus_trader_engine.services.kafka_ai_bridge import KafkaAIBridge
from nautilus_trader_engine.kafka_integration import KafkaConfig

router = APIRouter()

# Global instance of the KafkaAIBridge
# In a production environment, this would be managed as part of the app's lifecycle
kafka_config = KafkaConfig()
kafka_ai_bridge = KafkaAIBridge(kafka_config)

@router.on_event("startup")
async def startup_event():
    """Initializes the KafkaAIBridge on application startup."""
    if not await kafka_ai_bridge.initialize():
        raise RuntimeError("Failed to initialize KafkaAIBridge.")
    asyncio.create_task(kafka_ai_bridge.start())

@router.on_event("shutdown")
async def shutdown_event():
    """Shuts down the KafkaAIBridge on application shutdown."""
    await kafka_ai_bridge.stop()

@router.websocket("/ws/predictions")
async def subscribe_to_predictions(
    websocket: WebSocket,
    cache: CacheClient = Depends(get_cache_client)
):
    """
    WebSocket endpoint to stream real-time predictions with caching and confidence scoring.
    """
    await websocket.accept()
    prediction_uri = "ws://localhost:8765"  # The KafkaAIBridge WebSocket server

    try:
        async with websockets.connect(prediction_uri) as bridge_socket:
            logger.info("Successfully connected to the prediction bridge.")
            while True:
                prediction_json = await bridge_socket.recv()
                prediction_data = json.loads(prediction_json)

                # Validate prediction confidence
                if prediction_data.get("model_confidence", 0) < 0.6:
                    logger.debug(f"Skipping low-confidence prediction for {prediction_data['ticker']}")
                    continue

                # Cache the prediction
                cache_key = f"prediction:{prediction_data['ticker']}"
                await cache.setex(cache_key, 300, prediction_json)  # Cache for 5 minutes

                await websocket.send_text(prediction_json)

    except (WebSocketDisconnect, websockets.exceptions.ConnectionClosed) as e:
        logger.warning(f"Client or bridge connection closed: {e}")
    except Exception as e:
        logger.error(f"An error occurred in the prediction WebSocket: {e}", exc_info=True)
        await websocket.close(code=status.WS_1011_INTERNAL_ERROR)

@router.get("/history/{symbol}", summary="Get historical predictions for a symbol")
async def get_historical_predictions(
    symbol: str = Path(..., description="The stock symbol to get historical predictions for"),
    limit: int = Query(100, description="The number of historical predictions to return", ge=1, le=1000)
):
    """
    Retrieves a list of historical predictions for a given stock symbol.
    (This is a mock implementation and should be connected to a proper data store.)
    """
    # This should be implemented to fetch data from a time-series database or log files.
    mock_history = [
        {"timestamp": datetime.now(timezone.utc).isoformat(), "prediction": 150.0 + i * 0.1, "symbol": symbol}
        for i in range(limit)
    ]
    return {"status": "success", "data": mock_history}

@router.get("/accuracy", summary="Get model accuracy metrics")
async def get_model_accuracy():
    """
    Returns mock accuracy metrics for the prediction models.
    """
    mock_accuracy = {
        "1min_horizon": {"mean_absolute_error": 0.05, "accuracy": 0.85},
        "5min_horizon": {"mean_absolute_error": 0.12, "accuracy": 0.78},
        "15min_horizon": {"mean_absolute_error": 0.25, "accuracy": 0.72},
        "last_updated": datetime.now(timezone.utc).isoformat()
    }
    return {"status": "success", "data": mock_accuracy}